test/e2e_node/garbage_collector_test.go

```diff
@@ -0,0 +1,304 @@
+					restartCount:    3,
+					numContainers:   5,
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		// TODO (dashpole): Once the Container Runtime Interface (CRI) is complete, generalize run on other runtimes (other than docker)
+		dockerContainerGCTest(f, test)
+	}
+})
+
+// Tests the following:
+// 	pods are created, and all containers restart the specified number of times
+// 	while contianers are running, the number of copies of a single container does not exceed maxPerPodContainer
+// 	while containers are running, the total number of containers does not exceed maxTotalContainers
+// 	while containers are running, if not constrained by maxPerPodContainer or maxTotalContainers, keep an extra copy of each container
+// 	once pods are killed, all containers are eventually cleaned up
+func containerGCTest(f *framework.Framework, test testRun) {
+	Context(fmt.Sprintf("Garbage Collection Test: %s", test.testName), func() {
+		BeforeEach(func() {
+			realPods := getPods(test.testPods)
+			f.PodClient().CreateBatch(realPods)
+			By("Making sure all containers restart the specified number of times")
+			Eventually(func() error {
+				for _, podSpec := range test.testPods {
+					updatedPod, err := f.ClientSet.Core().Pods(f.Namespace.Name).Get(podSpec.podName)
+					if err != nil {
+						return err
+					}
+					if len(updatedPod.Status.ContainerStatuses) != podSpec.numContainers {
+						return fmt.Errorf("expected pod %s to have %d containers, actual: %d",
+							updatedPod.Name, podSpec.numContainers, len(updatedPod.Status.ContainerStatuses))
+					}
+					for _, containerStatus := range updatedPod.Status.ContainerStatuses {
+						if containerStatus.RestartCount != podSpec.restartCount {
+							return fmt.Errorf("pod %s had container with restartcount %d.  Should have been at least %d",
+								updatedPod.Name, containerStatus.RestartCount, podSpec.restartCount)
+						}
+					}
+				}
+				return nil
+			}, setupDuration, runtimePollInterval).Should(BeNil())
+		})
+
+		It(fmt.Sprintf("Should eventually garbage collect containers when we exceed the number of dead containers per container"), func() {
+			totalContainers := 0
+			for _, pod := range test.testPods {
+				totalContainers += pod.numContainers*2 + 1
+			}
+			Eventually(func() error {
+				total := 0
+				for _, pod := range test.testPods {
+					containerCount, err := pod.getContainerCount()
+					if err != nil {
+						return err
+					}
+					total += containerCount
+					// Check maxPerPodContainer
+					if containerCount > (maxPerPodContainer+1)*pod.numContainers {
+						return fmt.Errorf("expected total number of pod %v's containers: %v, to be <= (maxPerPodContainer+1)*numContainers %v",
+							pod.podName, containerCount, (maxPerPodContainer+1)*pod.numContainers)
+					}
+				}
+				//Check maxTotalContainers
+				if (maxTotalContainers > 0 || totalContainers <= maxTotalContainers) && total > maxTotalContainers {
```

Comment:
`maxTotalContainers > 0 || totalContainers <= maxTotalContainers` seems wrong to me.